File: Intersection of list.java
// Time complexity 0(m+n) (where m is the size of listA, n is the size of listB)
// Space complexity 0(n)
// code successfully executed in leetcode.

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA =0;
        int lenB = 0;
        ListNode current = headA;
        while(current!=null){
            current = current.next;
            lenA++;
        }
        current = headB;
        while(current!=null){
            current = current.next;
            lenB++;
        }
        while(lenA > lenB){
            headA = headA.next;
            lenA--;
        }
        while(lenB > lenA){
            headB = headB.next;
            lenB--;
        }
        while(headA != headB){
            headA = headA.next;
            headB = headB.next;
        }
        return headB;
        
    }
}

File: IntersectionofList.cs
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = 0;
        ListNode curr = headA;
        while(curr != null)
        {
            curr = curr.next;
            lenA++;
        }
        int lenB = 0;
        curr = headB;
        while(curr != null)
        {
            curr = curr.next;
            lenB++;
        }
        while(lenA > lenB)
        {
            headA = headA.next;
            lenA--;
        }
        while(lenB > lenA)
        {
            headB = headB.next;
            lenB--;
        }
        while(headA != headB)
        {
            headA = headA.next;
            headB = headB.next;
        }
        return headA;
    }
}
